File: viewer/server/schema.py
"""Strict validation for the only supported OmniX tensor schema."""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any, Mapping

import torch

from .config import IngestionLimits
from .errors import ConversionError, ResourceLimitError


REQUIRED_KEYS = frozenset(
    {"trajectory", "camera_pose", "intrinsics", "pts3d_dynamic_score"}
)


@dataclass(frozen=True, slots=True)
class ValidatedPredictions:
    trajectory: torch.Tensor
    camera_pose: torch.Tensor
    intrinsics: torch.Tensor
    dynamic_score: torch.Tensor

    @property
    def source_view_count(self) -> int:
        return int(self.trajectory.shape[0])

    @property
    def frame_count(self) -> int:
        return int(self.trajectory.shape[1])

    @property
    def height(self) -> int:
        return int(self.trajectory.shape[2])

    @property
    def width(self) -> int:
        return int(self.trajectory.shape[3])


def _safe_key(key: Any) -> str:
    if isinstance(key, str):
        return key[:80]
    return f"<{type(key).__name__}>"


def _validate_tensor_base(name: str, value: Any, rank: int) -> torch.Tensor:
    if not isinstance(value, torch.Tensor):
        raise ConversionError(
            "invalid_tensor_type",
            f"'{name}' must be a PyTorch tensor, not {type(value).__name__}.",
        )
    if value.layout != torch.strided or value.is_sparse or value.is_quantized:
        raise ConversionError(
            "unsupported_tensor_layout",
            f"'{name}' must be a dense, non-quantized strided tensor.",
        )
    if value.dtype != torch.float32:
        raise ConversionError(
            "unsupported_tensor_dtype",
            f"'{name}' must use float32 values.",
            details={"field": name, "received": str(value.dtype)},
        )
    if value.device.type != "cpu":
        raise ConversionError(
            "unsupported_tensor_device",
            f"'{name}' could not be normalized to CPU memory.",
        )
    if value.ndim != rank:
        raise ConversionError(
            "invalid_tensor_rank",
            f"'{name}' must have rank {rank}.",
            details={"field": name, "shape": list(value.shape)},
        )
    if any(int(dimension) <= 0 for dimension in value.shape):
        raise ConversionError(
            "invalid_tensor_shape",
            f"'{name}' must have only positive dimensions.",
            details={"field": name, "shape": list(value.shape)},
        )
    return value.detach()


def _all_finite(tensor: torch.Tensor, chunk_elements: int) -> bool:
    flat = tensor.reshape(-1)
    for start in range(0, flat.numel(), chunk_elements):
        if not bool(torch.isfinite(flat[start : start + chunk_elements]).all().item()):
            return False
    return True


def _rotation_determinants(rotation: torch.Tensor) -> torch.Tensor:
    a, b, c = rotation[:, 0, 0], rotation[:, 0, 1], rotation[:, 0, 2]
    d, e, f = rotation[:, 1, 0], rotation[:, 1, 1], rotation[:, 1, 2]
    g, h, i = rotation[:, 2, 0], rotation[:, 2, 1], rotation[:, 2, 2]
    return a * (e * i - f * h) - b * (d * i - f * g) + c * (d * h - e * g)


def validate_predictions(
    raw: Any, limits: IngestionLimits | None = None
) -> ValidatedPredictions:
    """Validate and normalize the exact plain-tensor OmniX output contract."""

    limits = limits or IngestionLimits()
    if type(raw) is not dict:
        raise ConversionError(
            "invalid_root_object",
            "The .pt root must be a plain dictionary containing four tensors.",
        )

    actual_keys = set(raw.keys())
    if actual_keys != REQUIRED_KEYS:
        missing = sorted(REQUIRED_KEYS - actual_keys, key=str)
        unexpected = sorted(
            (_safe_key(key) for key in actual_keys - REQUIRED_KEYS), key=str
        )[:16]
        raise ConversionError(
            "invalid_schema_keys",
            "The .pt keys do not match the supported OmniX output schema.",
            details={"missing": missing, "unexpected": unexpected},
        )

    trajectory = _validate_tensor_base("trajectory", raw["trajectory"], 5)
    camera_pose = _validate_tensor_base("camera_pose", raw["camera_pose"], 3)
    intrinsics = _validate_tensor_base("intrinsics", raw["intrinsics"], 3)
    dynamic_score = _validate_tensor_base(
        "pts3d_dynamic_score", raw["pts3d_dynamic_score"], 3
    )

    source_views, frames, height, width, xyz = map(int, trajectory.shape)
    if xyz != 3:
        raise ConversionError(
            "invalid_trajectory_shape",
            "'trajectory' must end in an xyz dimension of length 3.",
            details={"shape": list(trajectory.shape)},
        )
    expected_shapes = {
        "camera_pose": [source_views, 3, 4],
        "intrinsics": [source_views, 3, 3],
        "pts3d_dynamic_score": [source_views, height, width],
    }
    actual_shapes = {
        "camera_pose": list(camera_pose.shape),
        "intrinsics": list(intrinsics.shape),
        "pts3d_dynamic_score": list(dynamic_score.shape),
    }
    mismatches = {
        name: {"expected": expected_shapes[name], "received": actual_shapes[name]}
        for name in expected_shapes
        if expected_shapes[name] != actual_shapes[name]
    }
    if mismatches:
        raise ConversionError(
            "incompatible_tensor_shapes",
            "Camera and score shapes must match the trajectory source/image dimensions.",
            details={"mismatches": mismatches},
        )

    source_pixels = source_views * height * width
    if source_views > limits.max_source_views:
        raise ResourceLimitError(
            "source_view_limit_exceeded",
            f"The dataset has {source_views} source views; the limit is {limits.max_source_views}.",
        )
    if frames > limits.max_frames:
        raise ResourceLimitError(
            "frame_limit_exceeded",
            f"The dataset has {frames} frames; the limit is {limits.max_frames}.",
        )
    if source_pixels > limits.max_source_pixels:
        raise ResourceLimitError(
            "source_pixel_limit_exceeded",
            "The source-view pixel count exceeds the configured conversion limit.",
            details={"received": source_pixels, "limit": limits.max_source_pixels},
        )

    tensors = (trajectory, camera_pose, intrinsics, dynamic_score)
    logical_bytes = sum(tensor.numel() * tensor.element_size() for tensor in tensors)
    if logical_bytes > limits.max_total_tensor_bytes:
        raise ResourceLimitError(
            "tensor_byte_limit_exceeded",
            "The decoded tensor data exceeds the configured byte limit.",
            details={"received": logical_bytes, "limit": limits.max_total_tensor_bytes},
        )

    # Only normalize layout after all shape and byte ceilings have been checked.
    trajectory, camera_pose, intrinsics, dynamic_score = (
        tensor.contiguous() for tensor in tensors
    )
    for name, tensor in (
        ("trajectory", trajectory),
        ("camera_pose", camera_pose),
        ("intrinsics", intrinsics),
        ("pts3d_dynamic_score", dynamic_score),
    ):
        if not _all_finite(tensor, limits.finite_check_chunk_elements):
            raise ConversionError(
                "non_finite_tensor",
                f"'{name}' contains NaN or infinite values.",
                details={"field": name},
            )

    score_min = float(dynamic_score.min().item())
    score_max = float(dynamic_score.max().item())
    if score_min < -1e-6 or score_max > 1.0 + 1e-6:
        raise ConversionError(
            "invalid_dynamic_score_range",
            "'pts3d_dynamic_score' must contain probabilities in [0, 1].",
            details={"minimum": score_min, "maximum": score_max},
        )

    fx = intrinsics[:, 0, 0]
    fy = intrinsics[:, 1, 1]
    expected_bottom = torch.tensor([0.0, 0.0, 1.0], dtype=torch.float32)
    if bool((fx <= 0).any().item()) or bool((fy <= 0).any().item()):
        raise ConversionError(
            "invalid_intrinsics",
            "Camera focal lengths must be positive.",
        )
    if not bool(
        torch.isclose(intrinsics[:, 2, :], expected_bottom, atol=1e-4, rtol=1e-4)
        .all()
        .item()
    ):
        raise ConversionError(
            "invalid_intrinsics",
            "Each intrinsic matrix must have homogeneous row [0, 0, 1].",
        )

    rotation = camera_pose[:, :, :3]
    determinants = _rotation_determinants(rotation)
    gram = rotation.transpose(1, 2) @ rotation
    identity = torch.eye(3, dtype=torch.float32).expand_as(gram)
    if bool((determinants <= 0.5).any().item()) or bool(
        (determinants >= 1.5).any().item()
    ) or not bool(torch.isclose(gram, identity, atol=5e-2, rtol=5e-2).all().item()):
        raise ConversionError(
            "invalid_camera_pose",
            "Camera pose rotations must be non-singular, right-handed, and approximately orthonormal.",
        )

    return ValidatedPredictions(
        trajectory=trajectory,
        camera_pose=camera_pose,
        intrinsics=intrinsics,
        dynamic_score=dynamic_score,
    )

File: viewer/src/lib/pt/schema.ts
import type { TensorDescriptor } from './types'

const FLOAT32_BYTES = Float32Array.BYTES_PER_ELEMENT

export const OMX_MAX_SOURCE_VIEWS = 64
export const OMX_MAX_FRAMES = 600
export const OMX_MAX_SOURCE_PIXELS = 16_000_000
export const OMX_MAX_TENSOR_BYTES = 1024 * 1024 * 1024

const REQUIRED_KEYS = [
  'trajectory',
  'camera_pose',
  'intrinsics',
  'pts3d_dynamic_score',
] as const

type RequiredKey = (typeof REQUIRED_KEYS)[number]

export type OmnixSchemaErrorCode =
  | 'INVALID_SCHEMA_KEYS'
  | 'INVALID_TENSOR_DESCRIPTOR'
  | 'UNSUPPORTED_TENSOR'
  | 'INVALID_TENSOR_SHAPE'
  | 'RESOURCE_LIMIT_EXCEEDED'
  | 'NON_FINITE_TENSOR'
  | 'INVALID_DYNAMIC_SCORE'
  | 'INVALID_INTRINSICS'
  | 'INVALID_CAMERA_POSE'

export class OmnixSchemaError extends Error {
  readonly code: OmnixSchemaErrorCode

  constructor(code: OmnixSchemaErrorCode, message: string) {
    super(message)
    this.name = 'OmnixSchemaError'
    this.code = code
  }
}

export interface OmnixTensorSchema {
  trajectory: TensorDescriptor
  cameraPose: TensorDescriptor
  intrinsics: TensorDescriptor
  dynamicScore: TensorDescriptor
  sourceViewCount: number
  frameCount: number
  height: number
  width: number
  identitiesPerView: number
  identityCount: number
  totalTensorBytes: number
}

function fail(code: OmnixSchemaErrorCode, message: string): never {
  throw new OmnixSchemaError(code, message)
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
}

function checkedProduct(values: readonly number[], label: string): number {
  let result = 1
  for (const value of values) {
    if (!Number.isSafeInteger(value) || value <= 0) {
      fail('INVALID_TENSOR_SHAPE', `${label} must have positive, safe-integer dimensions.`)
    }
    if (result > Number.MAX_SAFE_INTEGER / value) {
      fail('RESOURCE_LIMIT_EXCEEDED', `${label} is too large for this browser.`)
    }
    result *= value
  }
  return result
}

function contiguousStride(shape: readonly number[]): number[] {
  const result = new Array<number>(shape.length)
  let stride = 1
  for (let index = shape.length - 1; index >= 0; index -= 1) {
    result[index] = stride
    stride *= shape[index]
  }
  return result
}

function assertTensorDescriptor(value: unknown, field: RequiredKey): TensorDescriptor {
  if (!isRecord(value) || !isRecord(value.storage)) {
    fail('INVALID_TENSOR_DESCRIPTOR', `'${field}' is not a supported tensor descriptor.`)
  }

  const descriptor = value as unknown as TensorDescriptor
  const storage = descriptor.storage
  if (
    storage.dtype !== 'float32' ||
    storage.location !== 'cpu' ||
    typeof storage.key !== 'string' ||
    storage.key.length === 0 ||
    !Number.isSafeInteger(storage.elementCount) ||
    storage.elementCount <= 0 ||
    !Number.isSafeInteger(storage.byteLength) ||
    storage.byteLength <= 0
  ) {
    fail('UNSUPPORTED_TENSOR', `'${field}' must use a CPU float32 storage.`)
  }
  if (
    !Number.isSafeInteger(descriptor.storageOffset) ||
    descriptor.storageOffset !== 0 ||
    !Array.isArray(descriptor.shape) ||
    !Array.isArray(descriptor.stride) ||
    descriptor.requiresGrad !== false
  ) {
    fail(
      'UNSUPPORTED_TENSOR',
      `'${field}' must be a detached, zero-offset, contiguous tensor.`,
    )
  }

  const elementCount = checkedProduct(descriptor.shape, `'${field}'`)
  if (descriptor.stride.length !== descriptor.shape.length) {
    fail('UNSUPPORTED_TENSOR', `'${field}' has an incompatible stride rank.`)
  }
  const expectedStride = contiguousStride(descriptor.shape)
  if (
    descriptor.stride.some(
      (stride, index) => !Number.isSafeInteger(stride) || stride !== expectedStride[index],
    )
  ) {
    fail('UNSUPPORTED_TENSOR', `'${field}' must use canonical contiguous strides.`)
  }
  if (
    storage.elementCount !== elementCount ||
    storage.byteLength !== elementCount * FLOAT32_BYTES
  ) {
    fail('INVALID_TENSOR_DESCRIPTOR', `'${field}' storage size does not match its shape.`)
  }
  return descriptor
}

function sameShape(actual: readonly number[], expected: readonly number[]): boolean {
  return actual.length === expected.length && actual.every((value, index) => value === expected[index])
}

/** Validate the exact plain-tensor dictionary produced by OmniX inference. */
export function validateOmnixTensorMap(
  raw: ReadonlyMap<string, TensorDescriptor>,
): OmnixTensorSchema {
  const keys = [...raw.keys()].sort()
  const expectedKeys = [...REQUIRED_KEYS].sort()
  if (
    keys.length !== expectedKeys.length ||
    keys.some((key, index) => key !== expectedKeys[index])
  ) {
    fail(
      'INVALID_SCHEMA_KEYS',
      'The .pt root must contain exactly the four supported OmniX prediction tensors.',
    )
  }

  const trajectory = assertTensorDescriptor(raw.get('trajectory'), 'trajectory')
  const cameraPose = assertTensorDescriptor(raw.get('camera_pose'), 'camera_pose')
  const intrinsics = assertTensorDescriptor(raw.get('intrinsics'), 'intrinsics')
  const dynamicScore = assertTensorDescriptor(
    raw.get('pts3d_dynamic_score'),
    'pts3d_dynamic_score',
  )

  if (trajectory.shape.length !== 5 || trajectory.shape[4] !== 3) {
    fail('INVALID_TENSOR_SHAPE', "'trajectory' must have shape [S, T, H, W, 3].")
  }
  const [sourceViewCount, frameCount, height, width] = trajectory.shape
  if (!sameShape(cameraPose.shape, [sourceViewCount, 3, 4])) {
    fail('INVALID_TENSOR_SHAPE', "'camera_pose' must have shape [S, 3, 4].")
  }
  if (!sameShape(intrinsics.shape, [sourceViewCount, 3, 3])) {
    fail('INVALID_TENSOR_SHAPE', "'intrinsics' must have shape [S, 3, 3].")
  }
  if (!sameShape(dynamicScore.shape, [sourceViewCount, height, width])) {
    fail(
      'INVALID_TENSOR_SHAPE',
      "'pts3d_dynamic_score' must have shape [S, H, W].",
    )
  }

  if (sourceViewCount > OMX_MAX_SOURCE_VIEWS) {
    fail('RESOURCE_LIMIT_EXCEEDED', 'The source-view count exceeds the browser limit.')
  }
  if (frameCount > OMX_MAX_FRAMES) {
    fail('RESOURCE_LIMIT_EXCEEDED', 'The trajectory frame count exceeds the browser limit.')
  }
  const identitiesPerView = checkedProduct([height, width], 'Source image')
  const identityCount = checkedProduct([sourceViewCount, height, width], 'Source pixels')
  if (identityCount > OMX_MAX_SOURCE_PIXELS) {
    fail('RESOURCE_LIMIT_EXCEEDED', 'The source pixel count exceeds the browser limit.')
  }

  const storageKeys = new Set([
    trajectory.storage.key,
    cameraPose.storage.key,
    intrinsics.storage.key,
    dynamicScore.storage.key,
  ])
  if (storageKeys.size !== REQUIRED_KEYS.length) {
    fail('UNSUPPORTED_TENSOR', 'The four OmniX tensors must use independent storages.')
  }

  const totalTensorBytes =
    trajectory.storage.byteLength +
    cameraPose.storage.byteLength +
    intrinsics.storage.byteLength +
    dynamicScore.storage.byteLength
  if (!Number.isSafeInteger(totalTensorBytes) || totalTensorBytes > OMX_MAX_TENSOR_BYTES) {
    fail('RESOURCE_LIMIT_EXCEEDED', 'The decoded tensor data exceeds the browser limit.')
  }

  return {
    trajectory,
    cameraPose,
    intrinsics,
    dynamicScore,
    sourceViewCount,
    frameCount,
    height,
    width,
    identitiesPerView,
    identityCount,
    totalTensorBytes,
  }
}

export function assertFiniteFloat32(name: string, values: Float32Array): void {
  for (let index = 0; index < values.length; index += 1) {
    if (!Number.isFinite(values[index])) {
      fail('NON_FINITE_TENSOR', `'${name}' contains NaN or infinite values.`)
    }
  }
}

export function validateDynamicScores(values: Float32Array): void {
  assertFiniteFloat32('pts3d_dynamic_score', values)
  for (let index = 0; index < values.length; index += 1) {
    if (values[index] < -1e-6 || values[index] > 1 + 1e-6) {
      fail('INVALID_DYNAMIC_SCORE', "'pts3d_dynamic_score' must contain values in [0, 1].")
    }
  }
}

function close(actual: number, expected: number, absolute = 1e-4, relative = 1e-4): boolean {
  return Math.abs(actual - expected) <= absolute + relative * Math.abs(expected)
}

/** Validate calibration tensors and return Three.js-basis camera matrices. */
export function validateAndConvertCameras(
  rawCameraPose: Float32Array,
  rawIntrinsics: Float32Array,
  sourceViewCount: number,
): { cameraPose: Float32Array; intrinsics: Float32Array } {
  if (
    rawCameraPose.length !== sourceViewCount * 12 ||
    rawIntrinsics.length !== sourceViewCount * 9
  ) {
    fail('INVALID_TENSOR_SHAPE', 'Camera tensor storage lengths do not match the schema.')
  }
  assertFiniteFloat32('camera_pose', rawCameraPose)
  assertFiniteFloat32('intrinsics', rawIntrinsics)

  const cameraPose = new Float32Array(sourceViewCount * 16)
  const intrinsics = new Float32Array(rawIntrinsics)
  const basis = [1, -1, -1, 1] as const

  for (let view = 0; view < sourceViewCount; view += 1) {
    const poseOffset = view * 12
    const intrinsicOffset = view * 9
    const fx = rawIntrinsics[intrinsicOffset]
    const fy = rawIntrinsics[intrinsicOffset + 4]
    if (
      fx <= 0 ||
      fy <= 0 ||
      !close(rawIntrinsics[intrinsicOffset + 6], 0) ||
      !close(rawIntrinsics[intrinsicOffset + 7], 0) ||
      !close(rawIntrinsics[intrinsicOffset + 8], 1)
    ) {
      fail('INVALID_INTRINSICS', 'Camera intrinsics are not valid homogeneous matrices.')
    }

    const a = rawCameraPose[poseOffset]
    const b = rawCameraPose[poseOffset + 1]
    const c = rawCameraPose[poseOffset + 2]
    const d = rawCameraPose[poseOffset + 4]
    const e = rawCameraPose[poseOffset + 5]
    const f = rawCameraPose[poseOffset + 6]
    const g = rawCameraPose[poseOffset + 8]
    const h = rawCameraPose[poseOffset + 9]
    const i = rawCameraPose[poseOffset + 10]
    const determinant = a * (e * i - f * h) - b * (d * i - f * g) + c * (d * h - e * g)
    if (determinant <= 0.5 || determinant >= 1.5) {
      fail('INVALID_CAMERA_POSE', 'Camera rotations must be right-handed and non-singular.')
    }
    const rotation = [a, b, c, d, e, f, g, h, i]
    for (let row = 0; row < 3; row += 1) {
      for (let column = 0; column < 3; column += 1) {
        let gram = 0
        for (let axis = 0; axis < 3; axis += 1) {
          gram += rotation[axis * 3 + row] * rotation[axis * 3 + column]
        }
        if (!close(gram, row === column ? 1 : 0, 5e-2, 5e-2)) {
          fail('INVALID_CAMERA_POSE', 'Camera rotations must be approximately orthonormal.')
        }
      }
    }

    const outputOffset = view * 16
    for (let row = 0; row < 4; row += 1) {
      for (let column = 0; column < 4; column += 1) {
        const raw = row < 3 ? (column < 4 ? rawCameraPose[poseOffset + row * 4 + column] : 0) : column === 3 ? 1 : 0
        cameraPose[outputOffset + row * 4 + column] = raw * basis[row] * basis[column]
      }
    }
  }

  return { cameraPose, intrinsics }
}
